### PR TITLE
add train, model, and related functions

### PR DESCRIPTION
**Pull request by kuanyutseng2**
add train, model, and related functions

### CODE REVIEW

File: src/ransacflow/model/resnet50.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+            return x
+        x = self.fc(x)
+        x = self.l2norm(x)
+
+        return x
+
+
+
+
+def resnet50(pretrained=False, **kwargs):
```
Comment (y3nr1ng) [inline src/ransacflow/model/resnet50.py]: @t5961p Can we switch over to `torchvision.models.resnet50` instead of rolling our own `ResNet` class?

File: src/ransacflow/model/model.py [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+        
+        return coef
+    
+    def forward(self, x, y):
+        
+        if self.training : 
+            coef = self.do_forward(x, y)
+        
+        else : ## if it is evaluation mode, switch to no grad, to accelerate 
+            with torch.no_grad() :
```
Comment (y3nr1ng) [inline src/ransacflow/model/model.py]: We don't do `no_grad` manually here, please move `do_forward` here instead. Set execution model to `train` or `eval` should be done before the train/evaluation loop.

File: src/ransacflow/model/model.py [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+            flow = torch.cat((flowX, flowY), dim=1)
+            flow = F.upsample_bilinear(flow, size=None, scale_factor=8) if up8X else flow
+        
+            return flow
+        
+        x = F.upsample_bilinear(x, size=None, scale_factor=8) if up8X else x
+        return x
+        
+        
+    def forward(self, x, up8X=True):
```
Comment (y3nr1ng) [inline src/ransacflow/model/model.py]: Similar issue as `CorrNeigh`, please move `do_forward` here.

File: src/ransacflow/model/model.py [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+    
+    def do_forward(self, x, y):
+        
+        ## x, y should be normalized
+        w, h = x.size()[2:]
+        coef = []
+        y = self.padding(y)
+        ## coef is the feature similarity between (i,j) and (i-r, j-r) with -kernel < r < +kernel 
+        for i,j in product(range(self.kernelSize), range(self.kernelSize)) : 
+            coef.append( torch.sum(x * y.narrow(2, i, w).narrow(3, j, h), dim=1, keepdim=True) )   # after sum: size = (n , 1 , w , h)
```
Comment (y3nr1ng) [inline src/ransacflow/model/model.py]: We can probably simplify this as a filter operation?

File: src/ransacflow/train.py [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+    finalGrad, final = model.predFlowCoarse(corr, network['netFlowCoarse'], grid)
+    
+    flowC = F.grid_sample(final[indexRoll].permute(0, 3, 1, 2), final).permute(0, 2, 3, 1)  # fill value from 1st argument[1] to the position 2nd argument[2,3]
+    ## cycle loss on flow
+    lossCycle = torch.mean(torch.abs(flowC - grid), dim=3).unsqueeze(1) ## Dim : N * 1 * W * H
+    lossCycle = torch.sum(lossCycle * maskMargin) / (torch.sum(maskMargin) + 0.001) 
+
+    ## Reconstruction Loss 3 channels 
+    IWarp = F.grid_sample(I, final)
+    lossLr =  LrLoss(IWarp, I[indexRoll], maskMargin, ssim)
```
Comment (y3nr1ng) [inline src/ransacflow/train.py]:
```
def SSIM(I1Warp, I2, match, ssim) :     
    return ssim(I1Warp, I2, match)
```
Instead of passing `ssim` this function as an argument, just directly send in `I1Warp`, `I2`, and `match` to that function. This occurs throughout this file.

File: src/ransacflow/train.py [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+            #a = torch.cuda.memory_allocated(0)
+            #c = torch.cuda.memory_reserved(0)
+            #print(f"memory allocated: {a}, reserved: {c}")
+
+            # Forward
+            for sub_optimizer in optimizer : 
+                sub_optimizer.zero_grad()
+                       
+             
+            lossLr, lossCycle, lossMatch, lossGrad, loss = LossFunction(network, I, indexRoll, grid, maskMargin, args, ssim, LrLoss)
```
Comment (y3nr1ng) [inline src/ransacflow/train.py]: Please refer to `computeLossMatchability` and `computeLossNoMatchability` function in this file. The complexity of these loss functions warrant their own class.

File: src/ransacflow/train.py [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+        # don't know why should times (1 - matchCycle)
+        #lossGrad = torch.sum(finalGrad * (1 - matchCycle[:, :, :-1, :-1]) * maskMargin[:, :, :-1, :-1]) / (torch.sum((1 - matchCycle[:, :, :-1, :-1]) * maskMargin[:, :, :-1, :-1]) + 0.001) 
+        lossGrad = torch.sum(finalGrad * (1 - matchCycle[:, :, :-1, :-1]) ) / (torch.sum((1 - matchCycle[:, :, :-1, :-1])) + 0.001) 
+        loss = loss + args.grad * lossGrad
+    
+
+    return lossLr.item(), lossCycle.item(), lossMatch.item(), lossGrad.item(), loss      
+
+## combine loss and gradloss together
+def computeLossNoMatchability(network, I, indexRoll, grid, maskMargin, args, ssim, LrLoss):
```
Comment (y3nr1ng) [inline src/ransacflow/train.py]: We can build a `torch.nn.module` for losses, for an example, `CrossEntropyLoss` has its own class inherit from `_Loss`. https://pytorch.org/docs/stable/_modules/torch/nn/modules/loss.html

File: src/ransacflow/train.py [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+    
+    f = F.normalize(network['netFeatCoarse'](I), p=2, dim=1) # feature extractor for source and target images
+    corr = network['netCorr'](f[indexRoll], f)   # source -> target & target -> source !?
+
+    finalGrad, final = model.predFlowCoarse(corr, network['netFlowCoarse'], grid)
```
Comment (y3nr1ng) [inline src/ransacflow/train.py]: Move the network we are training on outside of the loss function. We should only pass prediction results here, and focus entirely on 'evaluate the prediction results`.

File: src/ransacflow/train.py [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+            print (msg)
+
+    ## save network parameters in the last epoch
+    if df is not None :             
+        finalOut = os.path.join(args.outDir, 'BestModel@8_{:.3f}.pth'.format(bestPrec)) 
+        cmd = 'mv {} {}'.format(outNet, finalOut)
+        os.system(cmd)        
+
+
+if __name__ == '__main__':
```
Comment (y3nr1ng) [inline src/ransacflow/train.py]: I prefer to move these away, and do the following

1.  Create a script file, that wraps around the `train` function. This allows us to cleanly do (next one).
2.  Create a `.ipynb` that describes tasks with some preview along the way.

File: src/ransacflow/train.py [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+    bestPrec = 0
+    LastUpdate = 0
+
+    ## index of batch of images 
+    index = torch.arange(args.batchSize * 2).cuda()
+    indexRoll = torch.roll(index, args.batchSize).cuda()
+    
+
+    ###### Standard Training ######
+    for epoch in range(args.nEpochs):
```
Comment (y3nr1ng) [inline src/ransacflow/train.py]: I'm thinking about using `pytorch-lighning` instead of crafting our own training/evaluation loop.

**Review by y3nr1ng — CHANGES_REQUESTED** [review]
Some first thoughts after going through these skeleton codes.

File: src/ransacflow/train.py (hunk 5 above)
Comment (illyrs2) [reply src/ransacflow/train.py]: I changed the code to pass ssim directly. Currently working on changing to kornia but seems to fail.

**Review by y3nr1ng — APPROVED** [review]

**Review by y3nr1ng — APPROVED** [review]

File: src/ransacflow/train.py [hunk 11]
```diff
@@ -50,7 +50,7 @@ def computeLossMatchability(network, I, indexRoll, grid, maskMargin, args, ssim)
 
     ## Reconstruction Loss 3 channels 
     IWarp = F.grid_sample(I, final)
-    lossLr =  ssim(IWarp, I[indexRoll], matchCycle)
+    lossLr =  SSIM(IWarp, I[indexRoll], args.margin)
```
Comment (y3nr1ng) [inline src/ransacflow/train.py]: I think `margin` and `window` might be two different arguments?

Comment (illyrs2) [reply src/ransacflow/train.py]: The window size is fixed in def __init__(self, window_size = 11). 
And the margin is input as m (=88), where I crop the center of 48 by img1[..., m:-m, m:-m], img2[..., m:-m, m:-m] in the def forward () function.
